File: java/src/main/java/org/voidland/concurrent/turning_grille/WordsTrie.java
package org.voidland.concurrent.turning_grille;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.Locale;


public class WordsTrie
{
    private static class TrieNode
	{
		private TrieNode[] children;
		private boolean wordEnd;
		
		
		public TrieNode()
		{
			this.children = new TrieNode['Z' - 'A' + 1];
			this.wordEnd = false;
		}
		
		public TrieNode getOrCreateChild(char c)
		{
			int i = c - 'A';
			if (this.children[i] == null)
			{
				this.children[i] = new TrieNode();
			}
			return this.children[i];
		}
		
		public TrieNode getChild(char c)
		{
			int i = c - 'A';
			return this.children[i];
		}

		public boolean isWordEnd()
		{
			return this.wordEnd;
		}
		
		public void setWordEnd()
		{
			this.wordEnd = true;
		}
	}

    
    private TrieNode root;

	
	public WordsTrie(String wordsFilePath)
		throws IOException
	{
		this.root = new TrieNode();
		this.loadWords(wordsFilePath);
	}
	
    public int countWords(CharSequence text)
    {
        TrieNode[] iterators = new TrieNode[text.length() + 1];
        iterators[0] = this.root;
        
        int words = 0;
        for (int charIndex = 0; charIndex < text.length(); charIndex++)
        {
        	char c = text.charAt(charIndex);
            int i = 0;
            int j = 0;
            
            TrieNode currentNode;
      		while ((currentNode = iterators[i]) != null)
        	{
        		TrieNode nextIterator = currentNode.getChild(c);
        		iterators[i] = null;
        		++i;
        		
        		if (nextIterator == null)
        		{
        			continue;
        		}
        		
        		if (nextIterator.isWordEnd())
        		{
        			words++;
        		}
        		
        		iterators[j] = nextIterator;
        		++j;
        	}
    	    iterators[j] = this.root;
        }
    	return words;
    }

    private void loadWords(String wordsFilePath)
            throws IOException
    {
        try (BufferedReader wordsReader = new BufferedReader(new FileReader(wordsFilePath)))
        {
            String word;
            while ((word = wordsReader.readLine()) != null)
            {
                word = word.toUpperCase(Locale.ENGLISH);
                word = TurningGrilleCracker.NOT_ENGLISH_LETTERS_RE.matcher(word).replaceAll("");
                
                if (word.length() >= 3)	// Short words significantly increase the count of false positives.
                {
                	WordsTrie.addWord(this.root, word, 0);
                }
            }
        }
    }
    
    private static void addWord(TrieNode parent, CharSequence word, int charIndex)
    {
    	if (charIndex >= word.length())
    	{
    		parent.setWordEnd();
    		return;
    	}
    	TrieNode child = parent.getOrCreateChild(word.charAt(charIndex));
    	WordsTrie.addWord(child, word, charIndex + 1);
    }
}

File: rust/src/org/voidland/concurrent/turning_grille/words_trie.rs
use std::string::String;
use std::fs::read_to_string;
use core::str::Chars;
use super::NOT_CAPITAL_ENGLISH_LETTERS_RE;


const NUMBER_OF_LETTERS: usize = ('Z' as usize) - ('A' as usize) + 1;

struct TrieNode
{
    children: [Option<Box<Self>>; NUMBER_OF_LETTERS],
    wordEnd: bool
}

impl TrieNode
{
    #[inline]
    pub fn new() -> Self
    {
        Self
        {
            children: [const { None }; NUMBER_OF_LETTERS],
            wordEnd: false
        }
    }
    
    #[inline]
    pub fn getOrCreateChild<'a>(&'a mut self, c: char) -> &'a mut Self
    {
        let i: usize = (c as usize) - ('A' as usize);
        self.children[i].get_or_insert_with(|| Box::new(TrieNode::new()))
    }
    
    #[inline]
    pub fn getChild<'a>(&'a self, c: char) -> Option<&'a Self>
    {
        let i: usize = (c as usize) - ('A' as usize);
        let child: &Option<Box<Self>> = &self.children[i];
        match child
        {
            Some(child) => Some(child.as_ref()),
            None => None
        }
    }
    
    #[inline]
    pub fn isWordEnd(&self) -> bool
    {
        self.wordEnd
    }
    
    #[inline]
    pub fn setWordEnd(&mut self)
    {
        self.wordEnd = true;
    }
}


pub struct WordsTrie
{
	root: TrieNode
}

impl WordsTrie
{
    pub fn new(wordsFilePath: &str) -> Self
	{
		let mut ret: Self = Self
		{
			root: TrieNode::new()
		};
		ret.loadWords(wordsFilePath);
		ret
	}
    
    #[inline]
    pub fn countWords(&self, text: &str) -> usize
    {
        let mut iterators: Vec<Option<&TrieNode>> = vec![None; text.len() + 1];
        iterators[0] = Some(&self.root);
        
        let mut words: usize = 0;
        for c in text.chars()
        {
            let mut i: usize = 0;
            let mut j: usize = 0;
            
            while let Some(currentNode) = iterators[i]
            {
                let nextIterator: Option<&TrieNode> = currentNode.getChild(c);
                iterators[i] = None;
                i += 1;

                match nextIterator
                {
                    None =>
                    {
                        continue;
                    }
                    Some(nextNode) =>
                    {
                        if nextNode.isWordEnd()
                        {
                            words += 1;
                        }
                    }
                }
                iterators[j] = nextIterator;
                j += 1;
            }
            iterators[j] = Some(&self.root);
        }
        words
    }

    	fn loadWords(&mut self, wordsFilePath: &str)
	{
		for line in read_to_string(wordsFilePath).unwrap().lines()
		{
			let word: String = line.to_uppercase();
			let word = NOT_CAPITAL_ENGLISH_LETTERS_RE.replace_all(&word, "");
            
			if word.chars().count() >= 3
			{
				Self::addWord(&mut self.root, word.chars());
			}
		}
	}
	
	fn addWord(parent: &mut TrieNode, mut ci: Chars)
	{
		let c: Option<char> = ci.next();
        let child: &mut TrieNode = match c
        {
            None =>
            {
                parent.setWordEnd();
                return;
            }
            Some(c) =>
            {
                parent.getOrCreateChild(c)
            }
        };
		Self::addWord(child, ci);
	}
}
